File: risk-lib/src/test/java/net/mjduffin/risk/lib/entities/BoardTest.java
package net.mjduffin.risk.lib.entities;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class BoardTest {

    @Test
    void adjacentTerritoryTest() {
        Player player = new Player("TEST");
        Territory england = new Territory("England", player);
        Territory wales = new Territory("Wales", player);
        Board.Builder builder = new Board.Builder();
        List<Territory> territories = Arrays.asList(england, wales);
        builder.addTerritories(territories);
        builder.addEdge("England", "Wales");
        Board board = builder.build();

        assertTrue(board.areAdjacent(england, wales));
    }

    @Test
    void addPlayerTest() {
        //Test when adding a player to a territory that the player's unit count goes up by 1
        Player player = new Player("TEST");
        Territory england = new Territory("England", player);

        assertEquals(1, player.getTotalUnits());
        assertEquals(1, england.getUnits());
        assertEquals(0, england.getAvailableUnits());

    }

    @Test
    void increaseUnitsTest() {
        //Adding more units to a territory increases a player's units
        Player player = new Player("TEST");
        Territory england = new Territory("England", player);

        england.addUnits(5);

        assertEquals(6, player.getTotalUnits());
    }

}

File: risk-lib/src/test/kotlin/net/mjduffin/risk/lib/entities/BoardTest.kt
package net.mjduffin.risk.lib.entities

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

internal class BoardTest {

    @Test
    fun `adjacent territories should be detected`() {
        // given/when
        val england = TerritoryId("England")
        val wales = TerritoryId("Wales")
        val board = Board.Builder().addEdge(england, wales).build()

        // then
        assertTrue(board.areAdjacent(england, wales))
    }

    @Test
    fun `non-adjacent territories should be rejected`() {
        // given/when
        val england = TerritoryId("England")
        val wales = TerritoryId("Wales")
        val scotland = TerritoryId("Scotland")
        val board = Board.Builder()
            .addEdge(england, wales)
            .addEdge(england, scotland)
            .build()

        // then
        assertFalse(board.areAdjacent(wales, scotland))
    }


    @Test
    fun `connected territories should be detected`() {
        // given/when
        val england = TerritoryId("England")
        val wales = TerritoryId("Wales")
        val scotland = TerritoryId("Scotland")
        val board = Board.Builder()
            .addEdge(england, wales)
            .addEdge(england, scotland)
            .build()
        val a = PlayerId("PlayerA")
        val playerLookup = mapOf(Pair(england, a), Pair(wales, a), Pair(scotland, a))

        // then
        assertTrue(board.areConnected(wales, scotland, playerLookup))
    }


    @Test
    fun `unconnected territories should be rejected`() {
        // given/when
        val england = TerritoryId("England")
        val wales = TerritoryId("Wales")
        val scotland = TerritoryId("Scotland")
        val board = Board.Builder()
            .addEdge(england, wales)
            .addEdge(england, scotland)
            .build()
        val a = PlayerId("PlayerA")
        val b = PlayerId("PlayerB")
        // Wales (player A) <-> England (Player A) <-> Scotland (Player B)
        val playerLookup = mapOf(Pair(wales, b), Pair(england, a), Pair(scotland, b))

        // then
        assertFalse(board.areConnected(wales, scotland, playerLookup))
    }

    @Test
    fun `unconnected territories should be rejected (longer chain)`() {
        // given/when
        val england = TerritoryId("England")
        val wales = TerritoryId("Wales")
        val ireland = TerritoryId("Ireland")
        val scotland = TerritoryId("Scotland")
        val board = Board.Builder()
            .addEdge(ireland, wales)
            .addEdge(england, wales)
            .addEdge(england, scotland)
            .build()
        val a = PlayerId("PlayerA")
        val b = PlayerId("PlayerB")
        // Ireland (Player B) <-> Wales (player B) <-> England (Player A) <-> Scotland (Player B)
        val playerLookup = mapOf(Pair(ireland, b), Pair(wales, b), Pair(england, a), Pair(scotland, b))

        // then
        // Ireland is not connected to Scotland because England is occupied by player A
        assertFalse(board.areConnected(ireland, scotland, playerLookup))
    }

    @Test
    fun `full continents`() {
        // given/when
        val england = TerritoryId("England")
        val wales = TerritoryId("Wales")
        val scotland = TerritoryId("Scotland")
        val france = TerritoryId("France")
        val uk = Continent("UK", 3)
        val board = Board.Builder()
            .addEdge(england, wales)
            .addEdge(england, scotland)
            .addEdge(england, france)
            .addToContinent(uk, setOf(england, wales, scotland))
            .build()

        // then
        setOf(england, wales, scotland, )
        assertEquals(setOf(uk), board.fullContinents(setOf(england, wales, scotland, france)))
    }
}